grpc/stream.go

```diff
@@ -0,0 +1,325 @@
+func (s *stream) readData(wg *sync.WaitGroup) {
+	defer wg.Done()
+
+	for {
+		msg, err := s.stream.ReceiveConverted()
+
+		if err != nil && !isRegularClosing(err) {
+			s.logger.WithError(err).Debug("error while reading from the stream")
+
+			s.tq.Queue(func() error {
+				return s.closeWithError(err)
+			})
+
+			return
+		}
+
+		if len(msg) == 0 && isRegularClosing(err) {
+			s.logger.WithError(err).Debug("stream is cancelled/finished")
+
+			s.tq.Queue(func() error {
+				return s.closeWithError(nil)
+			})
+
+			return
+		}
+
+		s.queueMessage(msg)
+	}
+}
+
+func isRegularClosing(err error) bool {
+	return errors.Is(err, io.EOF) || errors.Is(err, grpcext.ErrCanceled)
+}
+
+// writeData writes data to the stream
+func (s *stream) writeData(wg *sync.WaitGroup) {
+	defer wg.Done()
+
+	writeChannel := make(chan message)
+
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+		for {
+			select {
+			case msg, ok := <-writeChannel:
+				if !ok {
+					return
+				}
+
+				err := s.stream.Send(msg.msg)
+				if err != nil {
+					s.logger.WithError(err).Error("failed to send data to the stream")
+
+					s.tq.Queue(func() error {
+						closeErr := s.closeWithError(err)
+						return closeErr
+					})
+					return
+				}
+			case _, ok := <-s.closeSend:
+				if !ok {
+					continue
+				}
+
+				if err := s.stream.CloseSend(); err != nil {
+					s.logger.WithError(err).Error("an error happened during stream closing")
+				}
+
+				close(s.closeSend)
+			case <-s.done:
+				return
+			}
+		}
+	}()
+
+	{
+		defer func() {
+			close(writeChannel)
+		}()
+
+		queue := make([]message, 0)
+		var wch chan message
+		var msg message
+
+		for {
+			wch = nil // this way if nothing to read it will just block
+			if len(queue) > 0 {
+				msg = queue[0]
+				wch = writeChannel
+			}
+			select {
+			case msg = <-s.writeQueueCh:
+				queue = append(queue, msg)
+			case wch <- msg:
+				queue = queue[:copy(queue, queue[1:])]
+
+			case <-s.done:
+				return
+			}
+		}
+	}
+}
+
+// on registers a listener for a certain event type
+func (s *stream) on(event string, listener func(goja.Value) (goja.Value, error)) {
+	if err := s.eventListeners.add(event, listener); err != nil {
+		s.vu.State().Logger.Warnf("can't register %s event handler: %s", event, err)
+	}
+}
+
+// write writes a message to the stream
+func (s *stream) write(input goja.Value) {
+	if input == nil || goja.IsNull(input) || goja.IsUndefined(input) {
+		s.logger.Warnf("can't send empty message")
+	}
+
+	rt := s.vu.Runtime()
+
+	b, err := input.ToObject(rt).MarshalJSON()
+	if err != nil {
+		s.logger.WithError(err).Warnf("can't marshal message")
+	}
+
+	s.writeQueueCh <- message{msg: b}
+}
+
+// end closes client the stream
+func (s *stream) end() {
+	var err error
+
+	defer func() {
+		_ = s.closeWithError(err)
+	}()
+
+	// send message only once
+	select {
+	case <-s.closeSend:
+		return
+	default:
+	}
+
+	s.closeSend <- struct{}{}
```

Comment:
Looking at this again - this is racy. 

THere is nothing stopping a call to `end()` to have finished for the select loop to be handling it and while that is happened as second call to `end` to get through the above select *before* the close of the channel and after it gets through the select, the close can happen and then we here try to write to it-> panic. 

So I guess we can just close it here?

I do prefer the usage of `state` field though as in websockets. Also I would argue that once the `stream` is closing, using it to write or end again should return an exception.

Reply:
I'm getting why you prefer the state over the channel since there will be a data race issue highlighting that the usage is wrong because using an event loop shouldn't cause such things.

So in the latest push, I've tried to switch to it. Let me know your thoughts on that.